src/core/ext/transport/chaotic_good/client_transport.cc

```diff
@@ -113,6 +118,70 @@ ClientTransport::ClientTransport(
       [](absl::Status status) {
         GPR_ASSERT(status.code() == absl::StatusCode::kCancelled ||
                    status.code() == absl::StatusCode::kInternal);
+        // TODO(ladynana): handle the promise endpoint write failures with
+        // outgoing_frames.close() once available.
+      });
+  auto read_loop = Loop([this] {
+    return TrySeq(
+        // Read frame header from control endpoint.
+        this->control_endpoint_->Read(FrameHeader::frame_header_size_),
+        // Read different parts of the server frame from control/data endpoints
+        // based on frame header.
+        [this](SliceBuffer read_buffer) mutable {
+          frame_header_ = std::make_shared<FrameHeader>(
+              FrameHeader::Parse(
+                  reinterpret_cast<const uint8_t*>(GRPC_SLICE_START_PTR(
+                      read_buffer.c_slice_buffer()->slices[0])))
+                  .value());
+          // Read header and trailers from control endpoint.
+          // Read message padding and message from data endpoint.
+          return TryJoin(
+              control_endpoint_->Read(frame_header_->GetFrameLength()),
+              TrySeq(
+                  data_endpoint_->Read(frame_header_->message_padding), [this] {
```

Comment:
Reading the padding first and then the message length may cause two separate syscalls which could be slow. Can we change this to read message_padding + message_length in one go and then later strip the padding away in the next promise possibly ?

Reply:
Sounds good, done.